ledger/participant-integration-api/src/main/scala/platform/apiserver/services/tracking/TrackerMap.scala

```diff
@@ -61,16 +61,36 @@ private[services] final class TrackerMap[Key](
       executionContext: ExecutionContext,
       loggingContext: LoggingContext,
   ): Future[Either[TrackedCompletionFailure, CompletionSuccess]] = {
+
+    def cleanClosed(
+        submitter: Key,
+        trackerResource: AsyncResource[TrackerWithLastSubmission],
+    ): Option[AsyncResource[TrackerWithLastSubmission]] = {
+      trackerResource.currentState match {
+        case Ready(tracker) if tracker.isCompleted =>
+          logger.info(
+            s"Removing completed tracker for $submitter"
+          )(trackerResource.loggingContext)
+          tracker.close()
+          trackerBySubmitter -= submitter
+          None
+        case Failed(_) | Closed =>
+          // simply forget already-failed or closed trackers
+          trackerBySubmitter -= submitter
+          None
+        case _ =>
+          Some(trackerResource)
+      }
+    }
+
     val key = getKey(submission.commands)
-    // double-checked locking
-    trackerBySubmitter
-      .getOrElse(
-        key,
-        lock.synchronized {
-          trackerBySubmitter.getOrElse(key, registerNewTracker(key))
-        },
-      )
-      .withResource(_.track(submission))
+    lock.synchronized {
```

Comment:
why change the locking approach?

Reply:
because I need to do something also on get

Reply:
as discussed offline this is OK